Cython/Compiler/ModuleNode.py

```diff
@@ -206,6 +218,14 @@ def analyse_declarations(self, env):
 
         self.body.analyse_declarations(env)
 
+        if self.find_shared_cy_lock_type(env):
+            # Be very suspicious of cython locks that are shared.
+            # They have the potential cause ABI issues.
+            self.scope.use_utility_code(
+                UtilityCode.load_cached(
+                    "CythonLockTypePublicCheck", "Lock.c"
+                ))
```

Comment:
I'm not completely sure this is the right place to do it. But it does seem to work.

It's currently undertested though